### ISSUE TITLE
Async searching is much slower than synchronous

### ISSUE THREAD

**Issue by drmikehenry** [mhinz/vim-grepper#174]
With Vim 8+, Grepper performs asynchronous searching.  When there are many
matches, asynchronous searching is much slower than synchronous searching, at
least when searching through a local tree of files on a fast hard drive.  For
example, using The Silver Searcher in a tree of around 2860 files having in
total 4658 matches for the string 'def', it takes about 10 seconds in Vim
8.1.290 on Ubuntu 16.04 to complete the search.  After setting the window
variable `testing` on the QuickFix window, to make Grepper perform the search
synchronously, it takes much less than one second (it's too fast to measure); at
the Bash prompt, measuring multiple iterations of this search yields a time of
about about 0.02 seconds per search.

Here is my test setup using a very minimal configuration.  First, start from an
empty Vim configuration:

```p sh
rm -rf ~/.vimrc ~/.vim
```

Then acquire vim-grepper and setup a small `vimrc` file:
```p sh

mkdir -p ~/.vim/bundle
git clone https://github.com/mhinz/vim-grepper.git ~/.vim/bundle/grepper

cat <<'EOF' > ~/.vim/vimrc
set nocompatible
set rtp^=$HOME/.vim/bundle/grepper
EOF
```

Launch Vim in a directory with many files having many matches, then run the
below command (or similar):

```p viml
:cgetexpr '' | Grepper -tool ag -query def
```

I'm not sure how much faster the asynchronous searching can be made, but for my
own searching needs, the results of synchronous searching with fast tools like
The Silver Searcher generally complete in well under a second.  Perhaps with
slower tools like Ack or on slow filesystems it's worth having asynchronous
search results, but at present I'm using Grepper with a patch that disables
asynchronous searching.  If it's not possible to fix asynchronous search speed
to be comparable to synchronous searching, it might be helpful to have a global
setting to disable asynchronous searching without needing to patch Grepper.

**Comment by lbonn** [comment]
See the related issue: #121.
The vim async api is the bottleneck here and this plugin is much faster with neovim.

**Comment by drmikehenry** [comment]
As a quick test to see how much faster Grepper can be made to work on Vim 8+, I
added the below patch.  Instead of using `caddexpr` for each line of stdout, it
just adds the line into an array.  The patch never adds these lines to the
QuickFix list; it's just a test to see how fast a plugin can respond to so many
lines of stdout.  It takes about one second to execute the search that before
was taking ten seconds.  Perhaps it would be possible to buffer lines of stdout
instead of adding them right away to the QuickFix list.  The lines could be
added in a batch either at the end of the operation, or when more than some
threshold of buffered lines is reached, or when there is a pause in receiving
lines of stdout for more than a certain amount of time.  This might lead to up
to a 10 factor speedup on Vim, since it seems that calling the `caddexpr`
function for each line of stdout seems to be the main bottleneck.

Here's the patch for speed testing:

    diff --git a/plugin/grepper.vim b/plugin/grepper.vim
    index ed3e20d..31b4e20 100644
    --- a/plugin/grepper.vim
    +++ b/plugin/grepper.vim
    @@ -188,7 +188,8 @@ function! s:on_stdout_vim(_job_id, data) dict abort
       let orig_dir = s:chdir_push(self.work_dir)
     
       try
    -    noautocmd execute self.addexpr 'a:data'
    +    "noautocmd execute self.addexpr 'a:data'
    +    let g:grepper_lines = g:grepper_lines + [a:data]
         if self.flags.stop > 0
               \ && len(self.flags.quickfix ? getqflist() : getloclist(0)) >= self.flags.stop
           call job_stop(s:id)
    @@ -746,6 +747,8 @@ function! s:run(flags)
             \ 'stdoutbuf': [],
             \ }
     
    +  let g:grepper_lines = []
    +
       call s:store_errorformat(a:flags)
     
       if &verbose

**Comment by drmikehenry** [comment]
I've submitted a pull request with a fix (#175).  It turns out that `cgetexpr` is not to blame for the slowness; instead, the repeated calls to `getqflist()` caused quadratic slowdowns.  The fix is only for Vim because I don't use Neovim, but I see that similar logic is used in the Neovim-specific portions of Grepper, and I suspect it would benefit from similar changes.

**Comment by lbonn** [comment]
Nice, we can probably close this now.
